@ndeloof Some failures (look related);

```
=== RUN   TestLinuxParseMountRaw
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:private"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> private  true  {volume name /absolute-path false  0x40000b1c68 0x40000b1c38 <nil> <nil>} false 0}
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:shared"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> shared  true  {volume name /absolute-path false  0x40000b1ce0 0x40000b1cb0 <nil> <nil>} false 0}
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:rslave"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> rslave  true  {volume name /absolute-path false  0x4000194060 0x4000194030 <nil> <nil>} false 0}
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:slave"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> slave  true  {volume name /absolute-path false  0x40001940a8 0x4000194078 <nil> <nil>} false 0}
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:rshared"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> rshared  true  {volume name /absolute-path false  0x40001942a0 0x4000194270 <nil> <nil>} false 0}
    linux_parser_test.go:91: ParseMountRaw(`"name:/absolute-path:rprivate"`) should have failed validation. Err '<nil>' - MP: &{ /absolute-path true name local volume <nil> rprivate  true  {volume name /absolute-path false  0x40001942e8 0x40001942b8 <nil> <nil>} false 0}
--- FAIL: TestLinuxParseMountRaw (0.00s)
```

```
=== RUN   TestValidateMount
    validate_test.go:57: expected "field VolumeOptions must not be specified", got "invalid mount config for type \"bind\": bind source path does not exist: /foo", case: 5
--- FAIL: TestValidateMount (0.00s)
```

Not sure if related (haven't looked at the test);

```
=== RUN   TestDockerSuite/TestRunNamedVolumesMountedAsShared
    docker_cli_run_test.go:3739: assertion failed: exitCode is 0
    --- FAIL: TestDockerSuite/TestRunNamedVolumesMountedAsShared (0.47s)
```

same:

```
=== RUN   TestDockerSuite/TestRunVolumeCopyFlag
    docker_cli_run_test.go:3872: assertion failed: expected an error, got nil
    --- FAIL: TestDockerSuite/TestRunVolumeCopyFlag (4.24s)
```

--- a/api/server/router/container/container_routes.go
+++ b/api/server/router/container/container_routes.go
@@ -16,6 +16,7 @@ import (
 	"github.com/docker/docker/api/types/backend"
 	"github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
 	"github.com/docker/docker/api/types/versions"
 	containerpkg "github.com/docker/docker/container"
 	"github.com/docker/docker/errdefs"
@@ -513,9 +514,42 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		}
 	}
 
+	if hostConfig != nil && versions.LessThan(version, "1.42") {
+		for _, m := range hostConfig.Mounts {
+			// Ignore BindOptions.CreateMountpoint because it was added in API 1.42.
+			if bo := m.BindOptions; bo != nil {
+				bo.CreateMountpoint = false
+			}
+
+			// These combinations are invalid, but weren't validated in API < 1.42.
+			// We reset them here, so that validation doesn't produce an error.
+			if o := m.VolumeOptions; o != nil && m.Type != mount.TypeVolume {
+				m.VolumeOptions = nil
+			}
+			if o := m.TmpfsOptions; o != nil && m.Type != mount.TypeTmpfs {
+				m.TmpfsOptions = nil
+			}
+			if bo := m.BindOptions; bo != nil {
+				// Ignore BindOptions.CreateMountpoint because it was added in API 1.42.
+				bo.CreateMountpoint = false
+			}
+		}
+	}
+
 	if hostConfig != nil && versions.GreaterThanOrEqualTo(version, "1.42") {
 		// Ignore KernelMemory removed in API 1.42.
 		hostConfig.KernelMemory = 0
+		for _, m := range hostConfig.Mounts {
+			if o := m.VolumeOptions; o != nil && m.Type != mount.TypeVolume {
+				return errdefs.InvalidParameter(fmt.Errorf("VolumeOptions must not be specified on mount type %q", m.Type))
+			}
+			if o := m.BindOptions; o != nil && m.Type != mount.TypeBind {
+				return errdefs.InvalidParameter(fmt.Errorf("BindOptions must not be specified on mount type %q", m.Type))
+			}
+			if o := m.TmpfsOptions; o != nil && m.Type != mount.TypeTmpfs {
+				return errdefs.InvalidParameter(fmt.Errorf("TmpfsOptions must not be specified on mount type %q", m.Type))
+			}
+		}
 	}
 
 	if hostConfig != nil && runtime.GOOS == "linux" && versions.LessThan(version, "1.42") {

--- a/api/swagger.yaml
+++ b/api/swagger.yaml
@@ -380,6 +380,10 @@ definitions:
             description: "Disable recursive bind mount."
             type: "boolean"
             default: false
+          CreateMountpoint:
+            description: "Create mount point on host if missing"
+            type: "boolean"
+            default: false
       VolumeOptions:
         description: "Optional configuration for the `volume` type."
         type: "object"

--- a/api/types/mount/mount.go
+++ b/api/types/mount/mount.go
@@ -82,8 +82,9 @@ const (
 
 // BindOptions defines options specific to mounts of type "bind".
 type BindOptions struct {
-	Propagation  Propagation `json:",omitempty"`
-	NonRecursive bool        `json:",omitempty"`
+	Propagation      Propagation `json:",omitempty"`
+	NonRecursive     bool        `json:",omitempty"`
+	CreateMountpoint bool        `json:",omitempty"`
 }
 
 // VolumeOptions represents the options for a mount of type volume.

--- a/daemon/volumes.go
+++ b/daemon/volumes.go
@@ -216,7 +216,7 @@ func (daemon *Daemon) registerMountPoints(container *container.Container, hostCo
 			}
 		}
 
-		if mp.Type == mounttypes.TypeBind {
+		if mp.Type == mounttypes.TypeBind && (cfg.BindOptions == nil || !cfg.BindOptions.CreateMountpoint) {
 			mp.SkipMountpointCreation = true
 		}
 

--- a/docs/api/version-history.md
+++ b/docs/api/version-history.md
@@ -97,6 +97,10 @@ keywords: "API, Docker, rcli, REST, documentation"
 * `POST /containers/create` on Linux now respects the `HostConfig.ConsoleSize` property.
   Container is immediately created with the desired terminal size and clients no longer
   need to set the desired size on their own.
+* `POST /containers/create` allow to set `CreateMountpoint` for host path to be
+  created if missing. This brings parity with `Binds`
+* `POST /containers/create` rejects request if BindOptions|VolumeOptions|TmpfsOptions
+  is set with a non-matching mount Type.
 
 ## v1.41 API changes
 

